### PR TITLE
Change: Make serde optional

### PR DESCRIPTION
**Pull request by devillove084**
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/342)

### CODE REVIEW

**Review by drmingdrmer — COMMENTED** [review]
Why is the `cfg(feature = "serde")` necessary in crate `memstore`?

File: memstore/src/lib.rs [hunk 1]
```diff
@@ -184,6 +185,7 @@ impl RaftLogReader<Config> for Arc<MemStore> {
     }
 }
 
+#[cfg(feature = "serde")]
```
[the remark this answers is not in the document]
Comment (devillove084) [reply memstore/src/lib.rs:188]: Because serdejson needs to be used in memstore, it must be enabled

[the remark this answers is not in the document]
Comment (devillove084) [reply memstore/src/lib.rs:188]: And cfg_attr is used in memstore to control the serialization switch of some classes.

File: memstore/src/lib.rs [hunk 2]
```diff
@@ -247,6 +249,7 @@ impl RaftSnapshotBuilder<Config, Cursor<Vec<u8>>> for Arc<MemStore> {
...
```
[the remark this answers is not in the document]
Comment (devillove084) [reply memstore/src/lib.rs:252]: Same as above

File: memstore/src/lib.rs [hunk 3]
```diff
@@ -359,6 +362,7 @@ impl RaftStorage<Config> for Arc<MemStore> {
         Ok(Box::new(Cursor::new(Vec::new())))
     }
 
+    #[cfg(feature = "serde")]
```
[the remark this answers is not in the document]
Comment (devillove084) [reply memstore/src/lib.rs:365]: Same as above

File: memstore/src/lib.rs [hunk 4]
```diff
@@ -421,6 +425,7 @@ impl RaftStorage<Config> for Arc<MemStore> {
     }
 
     type LogReader = Self;
+    #[cfg(feature = "serde")]
```
[the remark this answers is not in the document]
Comment (devillove084) [reply memstore/src/lib.rs:428]: Same as above

File: memstore/src/test.rs [hunk 5]
```diff
@@ -25,6 +25,7 @@ use crate::MemStore;
 ///     Suite::test_all(MemStoreBuilder {})
 /// }
 /// ```
+#[cfg(feature = "serde")]
```
[the remark this answers is not in the document]
Comment (devillove084) [reply memstore/src/test.rs:28]: Same as above

File: memstore/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (devillove084) [reply memstore/src/lib.rs:188]: Or I can remove all cfg here and change it to refer to openraft, but serde in memstore becomes non-optional

[the remark this answers is not in the document]
Comment (devillove084) [reply memstore/src/lib.rs]: Removed.

**Review by drmingdrmer — APPROVED** [review]
LGTM

Another thought is trait definitions does not need to be have to be `Serialize` and `Deserialize`. Because openraft itself does not depend on `serde`. What do you think? 🤔

**Pull request merged**